staging/src/k8s.io/apiserver/pkg/storageversion/updater.go

```diff
@@ -35,23 +35,96 @@ type Client interface {
+			sv.Status.CommonEncodingVersion = nil
+		}
+	}
+
+	condition := v1alpha1.StorageVersionCondition{
+		Type:               v1alpha1.AllEncodingVersionsEqual,
+		Status:             v1alpha1.ConditionFalse,
+		ObservedGeneration: sv.Generation,
+		LastTransitionTime: metav1.NewTime(time.Now()),
+		Reason:             "CommonEncodingVersionUnset",
+		Message:            "Common encoding version unset",
+	}
+	if sv.Status.CommonEncodingVersion != nil {
+		condition.Status = v1alpha1.ConditionTrue
+		condition.Reason = "CommonEncodingVersionSet"
+		condition.Message = "Common encoding version set"
+	}
+	forceTransition := false
+	if oldCommonEncodingVersion != nil && sv.Status.CommonEncodingVersion != nil &&
+		*oldCommonEncodingVersion != *sv.Status.CommonEncodingVersion {
+		forceTransition = true
+	}
+	setStatusCondition(&sv.Status.Conditions, condition, forceTransition)
+}
+
+func findStatusCondition(conditions []v1alpha1.StorageVersionCondition,
+	conditionType v1alpha1.StorageVersionConditionType) *v1alpha1.StorageVersionCondition {
+	for i := range conditions {
+		if conditions[i].Type == conditionType {
+			return &conditions[i]
+		}
+	}
+	return nil
+}
+
+// setStatusCondition sets the corresponding condition in conditions to newCondition.
+// conditions must be non-nil.
+// 1. if the condition of the specified type already exists (all fields of the existing condition are updated to
+//    newCondition, LastTransitionTime is set to now if the new status differs from the old status)
+// 2. if a condition of the specified type does not exist (LastTransitionTime is set to now() if unset, and newCondition is appended)
+// NOTE: forceTransition allows overwriting LastTransitionTime even when the status doesn't change.
+func setStatusCondition(conditions *[]v1alpha1.StorageVersionCondition, newCondition v1alpha1.StorageVersionCondition,
+	forceTransition bool) {
+	if conditions == nil {
 		return
 	}
-	firstVersion := sv.Status.StorageVersions[0].EncodingVersion
-	agreed := true
-	for _, ssv := range sv.Status.StorageVersions {
-		if ssv.EncodingVersion != firstVersion {
-			agreed = false
-			break
+	existingCondition := findStatusCondition(*conditions, newCondition.Type)
+	if existingCondition == nil {
+		if newCondition.LastTransitionTime.IsZero() {
+			newCondition.LastTransitionTime = metav1.NewTime(time.Now())
 		}
+		*conditions = append(*conditions, newCondition)
+		return
 	}
-	if agreed {
-		sv.Status.CommonEncodingVersion = &firstVersion
-	} else {
-		sv.Status.CommonEncodingVersion = nil
+
+	if existingCondition.Status != newCondition.Status || forceTransition {
+		existingCondition.Status = newCondition.Status
+		if !newCondition.LastTransitionTime.IsZero() {
+			existingCondition.LastTransitionTime = newCondition.LastTransitionTime
+		} else {
+			existingCondition.LastTransitionTime = metav1.NewTime(time.Now())
+		}
 	}
+
+	existingCondition.Reason = newCondition.Reason
+	existingCondition.Message = newCondition.Message
```

Comment:
this is the bit that seems likely to duel... should we skip if the ObservedGeneration of the existing condition is already set to the current generation?

Reply:
StorageVersion has no spec, so the ObservedGeneration is always 0. 

IIUC, https://github.com/kubernetes/kubernetes/pull/78133 happened because one controller's update would always trigger another controller to do calculation and update. This doesn't apply to storage version because:
- The GC only tries updating status if the live lookup object has stale record. The watch event of one controller finishing GC won't trigger another controller to do GC again.
- The StorageVersion manager doesn't watch.

Reply:
One storageVersion GC or on API server updating the condition of a storageVersion won't cause other GC or API server update the storageVersion again, so this won't cause a duel.